### PR TITLE
Onboarding step 3 - weather api service, home/details screens integration

### PR DESCRIPTION
**Pull request by s-zak**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -7,17 +7,22 @@
...
+    "@sinclair/typebox": "^0.31.23",
+    "axios": "^1.6.1",
     "react": "18.2.0",
     "react-native": "0.72.6",
+    "react-native-config": "^1.5.1",
     "react-native-gesture-handler": "^2.13.4",
     "react-native-safe-area-context": "^4.7.4",
-    "react-native-screens": "^3.27.0"
+    "react-native-screens": "^3.27.0",
+    "react-query": "^3.39.3"
```
Comment (jaworek) [inline package.json:25]: FYI, there's a newer version of RQ that uses different package name: `@tanstack/react-query`

File: src/screens/homeScreen/hooks/useWeatherForCities.tsx [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  });
+
+  const refetch = useRefCallback(() => {
+    if (isFetching) {
+      return;
+    }
+    _refetch({throwOnError: true});
+  });
+
+  useFocusEffect(refetch);
```
Comment (jaworek) [inline src/screens/homeScreen/hooks/useWeatherForCities.tsx:44]: Not sure if it is available in older version of RQ, but this is how they recommend handling it in their docs:
https://tanstack.com/query/latest/docs/react/react-native#refetch-on-app-focus

File: src/screens/homeScreen/hooks/useWeatherForCities.tsx [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import constants from 'app/constants';
+import {useRefCallback} from 'app/hooks/useRefCallback';
+import {CityListItem} from 'app/screens/homeScreen/components/CItyWeatherList';
+
+type UseWeatherForCities = {
+  isLoading: boolean;
+  data?: CityListItem[];
+  error?: APIError | Error | null;
+  refetch: () => void;
+};
```
Comment (jaworek) [inline src/screens/homeScreen/hooks/useWeatherForCities.tsx:16]: I would avoid typing return type as it requires additional maintenance and also can result in worse types.

Explanation from Matt and Theo:
https://www.youtube.com/watch?v=nwSe95uFN8E
https://www.youtube.com/watch?v=I6V2FkW1ozQ

And a bit different take from Prime:
https://www.youtube.com/watch?v=Gcr4t6cH-lU

File: src/screens/homeScreen/HomeScreen.tsx [hunk 4]
```diff
@@ -1,21 +1,79 @@
...
+import colors from 'app/styles/colors';
+
+import {useWeatherForCities} from './hooks/useWeatherForCities';
 
 function HomeScreen(): JSX.Element {
   const navigation = useNavigation();
+  const insets = useSafeAreaInsets();
+  const {isLoading, data, error, refetch} = useWeatherForCities();
+
+  const {listContentContainerStyle, centeredWithBottomInset} = useMemo<{
```
Comment (jaworek) [inline src/screens/homeScreen/HomeScreen.tsx:28]: Having a `useMemo` here seems a bit like a premature optimisation. Creating this type of object does not seem to be computationally intensive. Unless, there's a measured difference in performance, I would prefer not using it.

https://kentcdodds.com/blog/usememo-and-usecallback

File: package.json (hunk 1 above)
Comment (s-zak) [reply package.json:25]: Upgraded to v5 - https://github.com/callstack-internal/szymon-zakrzewski-onboarding/pull/5/commits/bb0c67038635a6b82da6ad38a6737437318c6507

File: src/screens/homeScreen/hooks/useWeatherForCities.tsx (hunk 2 above)
Comment (s-zak) [reply src/screens/homeScreen/hooks/useWeatherForCities.tsx:44]: I've checked it and it looks that both solutions do more or less the same job, but my approach also prevents excessive data refetching, i.g. when user moves quickly between screens. So I'll leave it as is.

File: src/screens/homeScreen/hooks/useWeatherForCities.tsx (hunk 3 above)
Comment (s-zak) [reply src/screens/homeScreen/hooks/useWeatherForCities.tsx:16]: Great materials, thx! After watching it I totally agree to remove these typings.

File: src/screens/homeScreen/HomeScreen.tsx (hunk 4 above)
Comment (s-zak) [reply src/screens/homeScreen/HomeScreen.tsx:28]: True, adding it here does't make sense - fixed in https://github.com/callstack-internal/szymon-zakrzewski-onboarding/pull/5/commits/a7bd66af55921679fa9f1eefac1521c210a2e080

**Review by jaworek — APPROVED** [review]

**Pull request merged**